services/wallet/storage/storage_test.go

```diff
@@ -0,0 +1,318 @@
+			ctx := context.Background()
+
+			_, err1 := dbi.ExecContext(ctx, q, tc.given.chal.ID, tc.given.chal.CreatedAt, tc.given.chal.Nonce)
+			must.NoError(t, err1)
+
+			c := Challenge{}
+			actual, err2 := c.Get(ctx, dbi, tc.given.id)
+			must.Equal(t, tc.exp.err, err2)
+
+			must.Equal(t, tc.exp.chal.ID, actual.ID)
+			must.Equal(t, tc.exp.chal.CreatedAt, actual.CreatedAt)
+			must.Equal(t, tc.exp.chal.Nonce, actual.Nonce)
+		})
+	}
+}
+
+func TestChallenge_Upsert(t *testing.T) {
+	dbi, err := setupDBI()
+	must.NoError(t, err)
+
+	defer func() {
+		_, _ = dbi.Exec("TRUNCATE_TABLE challenge;")
+	}()
+
+	const q = `select * from challenge where id = $1`
+
+	chlRepo := &Challenge{}
+
+	t.Run("insert", func(t *testing.T) {
+		exp := model.Challenge{
+			ID:        "1",
+			CreatedAt: time.Date(2024, 1, 1, 1, 1, 1, 0, time.UTC),
+			Nonce:     "a",
+		}
+
+		err1 := chlRepo.Upsert(context.TODO(), dbi, exp)
+		must.NoError(t, err1)
+
+		var actual model.Challenge
+		err2 := sqlx.GetContext(context.TODO(), dbi, &actual, q, exp.ID)
+		must.NoError(t, err2)
+
+		must.Equal(t, exp.ID, actual.ID)
+		must.Equal(t, exp.CreatedAt, actual.CreatedAt)
+		must.Equal(t, exp.Nonce, actual.Nonce)
+	})
+
+	t.Run("upsert", func(t *testing.T) {
+		ctx := context.Background()
+
+		exp := model.Challenge{
+			ID:        "1",
+			CreatedAt: time.Date(2024, 12, 1, 1, 1, 1, 0, time.UTC),
+			Nonce:     "b",
+		}
+
+		err1 := chlRepo.Upsert(ctx, dbi, exp)
+		must.NoError(t, err1)
+
+		var actual model.Challenge
+		err2 := sqlx.GetContext(ctx, dbi, &actual, q, exp.ID)
+		must.NoError(t, err2)
+
+		must.Equal(t, exp.ID, actual.ID)
+		must.Equal(t, exp.CreatedAt, actual.CreatedAt)
+		must.Equal(t, exp.Nonce, actual.Nonce)
+	})
+}
+
+func TestChallenge_Delete(t *testing.T) {
+	dbi, err := setupDBI()
+	must.NoError(t, err)
+
+	defer func() {
+		_, _ = dbi.Exec("TRUNCATE_TABLE challenge;")
+	}()
+
+	type tcGiven struct {
+		id   string
+		chal model.Challenge
+	}
+
+	type exp struct {
+		err error
+	}
+
+	type testCase struct {
+		name  string
+		given tcGiven
+		exp   exp
+	}
+
+	tests := []testCase{
+		{
+			name: "delete",
+			given: tcGiven{
+				id: "1",
+				chal: model.Challenge{
+					ID:    "1",
+					Nonce: "nonce-1",
+				}},
+			exp: exp{
+				err: nil,
+			},
+		},
+		{
+			name: "no_rows_deleted",
+			given: tcGiven{
+				id: "some-random-id",
+				chal: model.Challenge{
+					ID:        "2",
+					CreatedAt: time.Date(2024, 1, 1, 1, 1, 1, 0, time.UTC),
+					Nonce:     "nonce-2",
+				}},
+			exp: exp{
+				err: model.ErrNoRowsDeleted,
+			},
+		},
+	}
+
+	const q = `insert into challenge (id, created_at, nonce) values($1, $2, $3)`
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := context.Background()
+
+			_, err1 := dbi.ExecContext(ctx, q, tc.given.chal.ID, tc.given.chal.CreatedAt, tc.given.chal.Nonce)
+			must.NoError(t, err1)
+
+			c := Challenge{}
+			err2 := c.Delete(ctx, dbi, tc.given.id)
+			must.Equal(t, tc.exp.err, err2)
```

Comment:
This error here acts as data we're checking for, therefore:

```suggestion
			should.Equal(t, tc.exp.err, err2)
```